openedx_events/event_bus/avro/schema.py

```diff
@@ -0,0 +1,95 @@
+"""
+Code to convert attr classes to Avro specification.
+
+TODO (EventBus): handle optional parameters and allow for schema evolution (ARCHBOM-2013)
+"""
+from datetime import datetime
+
+from opaque_keys.edx.keys import CourseKey
+
+from .custom_serializers import CourseKeyAvroSerializer, DatetimeAvroSerializer
+from .types import PYTHON_TYPE_TO_AVRO_MAPPING
+
+DEFAULT_FIELD_TYPES = {
+    datetime: DatetimeAvroSerializer.field_type,
+    CourseKey: CourseKeyAvroSerializer.field_type,
+}
+
+
+def schema_from_signal(signal, custom_field_types=None):
+    """Create an Avro schema for events sent by an instance of OpenEdxPublicSignal."""
+    field_types = custom_field_types or {}
+    all_custom_field_types = {**DEFAULT_FIELD_TYPES, **field_types}
+    schema_record_names = set()
+
+    base_schema = {
+        "name": "CloudEvent",
+        "type": "record",
+        "doc": "Avro Event Format for CloudEvents created with openedx_events/schema",
+        "fields": [],
+    }
+
+    for data_key, data_type in signal.init_data.items():
+        base_schema["fields"].append(_create_avro_field_definition(data_key, data_type,
+                                                                   schema_record_names,
+                                                                   custom_field_types=all_custom_field_types))
+    return base_schema
+
+
+def _create_avro_field_definition(data_key, data_type, schema_record_names,
+                                  custom_field_types=None, default_is_none=False):
+    """
+    Create an Avro schema field definition from an OpenEdxPublicSignal data definition.
+
+    Arguments:
+        data_key: Field name
+        data_type: Python data type, eg `str`, `CourseKey`, `CourseEnrollmentData`
+        schema_record_names: list of previously-encountered data types
```

Comment:
If these are data types as noted in the string, what about a name like `processed_data_types`?